fixbug(router.py): add original_messages handling in Router class

FatGhosta opened the pull request:
to fixbug in https://github.com/BerriAI/litellm/issues/9816
Introduces original_messages to store a deep copy of input messages. This change allows for better management of message states during fallback operations, ensuring that the original messages are preserved and can be reused when necessary.

I have already conducted tests in my own project. I'm sorry that due to certain reasons, I can't access OpenAI or Gemini locally.

## Title

<!-- e.g. "Implement user authentication feature" -->

## Relevant issues

<!-- e.g. "Fixes #000" -->

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on (`make test-unit`)[https://docs.litellm.ai/docs/extras/contributing_code]
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

<!-- Select the type of Pull Request -->
<!-- Keep only the necessary ones -->

🐛 Bug Fix

## Changes

FatGhosta committed:
fixbug(router.py): add original_messages handling in Router class

Introduces original_messages to store a deep copy of input messages.
This change allows for better management of message states during
fallback operations, ensuring that the original messages are preserved
and can be reused when necessary.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/BerriAI/litellm?pullRequest=9912) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/BerriAI/litellm?pullRequest=9912) before we can accept your contribution.<br/><hr/>**zhangjie** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/BerriAI/litellm?pullRequest=9912) it.</sub>

krrishdholakia commented on litellm/router.py:
```diff
@@ -3144,6 +3144,7 @@ async def async_function_with_fallbacks(self, *args, **kwargs):  # noqa: PLR0915
         model_group: Optional[str] = kwargs.get("model")
         disable_fallbacks: Optional[bool] = kwargs.pop("disable_fallbacks", False)
         fallbacks: Optional[List] = kwargs.get("fallbacks", self.fallbacks)
+        original_messages: Optional[List] = copy.deepcopy(kwargs.get("messages", None))
```
Hi @FatGhosta can you please add a test to https://github.com/BerriAI/litellm/blob/fb9e5dbd6ed71fc9eea86950c9da149f24e6a16c/tests/litellm/test_router.py#L4

to confirm your fix solves the problem

FatGhosta replied:
My test didn't pass. It seems that it fail to get environment variable such as the API key in the testing environment. Do you have any suggestions?

FatGhosta replied:
@krrishdholakia 
I introduced the MagicMock component this time to mock the client, and then verified the parameter changes in the request call chain:
For the first request to Gemini, the image_url will be converted to base64 encoding, which is different from the original request;
For the second fallback request to GPT-4o, the original_message will be replaced back into the kwargs parameter before the request.

FatGhosta committed:
feat(tests/litellm/test_router.py): add test for router with image_url

Introduces a new test case to validate the router's behavior when
handling image URLs, particularly in fallback scenarios. This
enhancement ensures that the router correctly processes image
URLs and utilizes fallbacks as expected, improving overall
test coverage and reliability.

For details, refer to issue: https://github.com/BerriAI/litellm/issues/9816

FatGhosta committed:
feat(router.py): enhance original_messages handling in Router

Improves the handling of original_messages in the Router class by
clearing and deep copying messages when they are provided. This
ensures that the Router can effectively manage message states during
fallback operations, enhancing the overall functionality and reliability
of the routing process.

Additionally, updates the test for router behavior with image URLs to
validate the new message handling logic, ensuring that the router
correctly processes and modifies messages as expected.
